kernel/src/sched.rs

```diff
@@ -50,11 +130,35 @@ pub struct Kernel {
     grants_finalized: Cell<bool>,
 }
 
+/// Enum used to inform scheduler why a process stopped
+/// executing (aka why `do_process()` returned).
+#[derive(PartialEq, Eq)]
+pub enum StoppedExecutingReason {
+    /// The process returned because it is no longer ready to run
+    NoWorkLeft,
+
+    /// The process faulted, and the board restart policy was configured such that
+    /// it was not restarted and there was not a kernel panic.
+    StoppedFaulted,
+
+    /// The kernel stopped the process
+    Stopped,
+
+    /// The process returned because its timeslice expired
+    TimesliceExpired,
+
+    /// The process returned because it was preempted by kernel
+    /// work that became ready (most likely because an interrupt fired
+    /// and the kernel thread needs to execute the bottom half of the
+    /// interrupt)
```

Comment:
```suggestion
    /// The process returned because it was preempted by the kernel.
    /// This can mean that kernel work became ready (most likely because an interrupt fired
    /// and the kernel thread needs to execute the bottom half of the
    /// interrupt), or because the scheduler no longer wants to execute that
    /// process.
```